### PR TITLE
feat(build,docker,buildah): add secrets support for docker and buildah

### PR DESCRIPTION
**Pull request by iapershin**
This add new feature "build secrets" for docker and buildah

### CODE REVIEW

File: pkg/config/raw_image_from_dockerfile.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ type rawImageFromDockerfile struct {
 	Staged          bool                   `yaml:"staged,omitempty"`
 	Platform        []string               `yaml:"platform,omitempty"`
```
Comment (alexey-igrychev) [inline pkg/config/raw_image_from_dockerfile.go:25]: unnecessary line break

File: pkg/config/raw_image_from_dockerfile.go [hunk 2]
```diff
@@ -142,6 +144,29 @@ func (c *rawImageFromDockerfile) toImageFromDockerfileDirective(giterminismManag
 	image.platform = append([]string{}, c.Platform...)
 	image.raw = c
 
+	secretIds := make(map[string]struct{})
+	for _, rawSecrets := range c.RawSecrets {
+
```
Comment (alexey-igrychev) [inline pkg/config/raw_image_from_dockerfile.go:149]: unnecessary line break

File: test/e2e/build/_fixtures/simple/state1/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,11 @@
+FROM ubuntu:22.04
+
+RUN --mount=type=secret,id=ENV_SECRET cat /run/secrets/ENV_SECRET
```
Comment (alexey-igrychev) [inline test/e2e/build/_fixtures/simple/state1/Dockerfile:3]: Let's also check the secret value. Something like `[ "$(cat /run/secrets/ENV_SECRET)" = "value" ]`.

Comment (alexey-igrychev) [reply test/e2e/build/_fixtures/simple/state1/Dockerfile:3]: Let's hardcode it.

File: pkg/container_backend/interface.go [hunk 4]
```diff
@@ -41,6 +41,8 @@ type BuildDockerfileOpts struct {
 	SSH                  string
 	Labels               []string
 	Tags                 []string
+
```
Comment (alexey-igrychev) [inline pkg/container_backend/interface.go:44]: unnecessary line break

File: pkg/config/raw_secrets.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		return fmt.Errorf("secrets validation error: %w", err)
+	}
+
+	return nil
+}
+
+func (s *rawSecret) validate() error {
+	switch {
+	case s.Env != "":
+		if s.Src != "" || s.PlainValue != "" {
```
Comment (alexey-igrychev) [inline pkg/config/raw_secrets.go:42]: You can use something like that https://github.com/werf/werf/blob/e15cf3ce7ce16de984856e651e87067367e55b3e/pkg/config/common.go#L65 to check it in one place.

Comment (alexey-igrychev) [reply pkg/config/raw_secrets.go:42]: https://github.com/werf/werf/blob/e15cf3ce7ce16de984856e651e87067367e55b3e/pkg/config/git_remote_export.go#L26-L28

File: pkg/config/secrets.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+func (s *SecretFromPlainValue) GetSecretStringArg() (string, error) {
+	secret, err := s.setPalinValueAsEnv()
+	if err != nil {
+		return "", err
+	}
+	return secret.GetSecretStringArg()
+}
+
+func (s *SecretFromPlainValue) setPalinValueAsEnv() (*SecretFromEnv, error) {
```
Comment (alexey-igrychev) [inline pkg/config/secrets.go:52]:
```suggestion
func (s *SecretFromPlainValue) setPlainValueAsEnv() (*SecretFromEnv, error) {
```

File: pkg/config/secrets.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	return &SecretFromEnv{
+		Id:    s.Id,
+		Value: envKey,
+	}, nil
+}
+
+func (s *SecretFromEnv) GetSecretId() string {
+	return s.Id
```
Comment (alexey-igrychev) [inline pkg/config/secrets.go:71]: What do you think about moving the ID logic here instead of within casting Raw to non-Raw? It seems more clear.

**Comment by alexey-igrychev** [comment]
+ let's fail when staged Dockerfile used

File: pkg/config/image_from_dockerfile.go [hunk 8]
```diff
@@ -21,6 +21,8 @@ type ImageFromDockerfile struct {
 	Dependencies    []*Dependency
 	Staged          bool
```
[the remark this answers is not in the document]
Comment (iapershin) [reply pkg/config/image_from_dockerfile.go]: deleted

File: test/e2e/build/_fixtures/simple/state1/Dockerfile (hunk 3 above)
Comment (iapershin) [reply test/e2e/build/_fixtures/simple/state1/Dockerfile]: done

File: pkg/container_backend/interface.go (hunk 4 above)
Comment (iapershin) [reply pkg/container_backend/interface.go]: deleted

File: pkg/config/raw_secrets.go (hunk 5 above)
Comment (iapershin) [reply pkg/config/raw_secrets.go]: done

File: pkg/config/secrets.go (hunk 6 above)
Comment (iapershin) [reply pkg/config/secrets.go]: awesome catch. fixed

File: pkg/config/secrets.go (hunk 7 above)
Comment (iapershin) [reply pkg/config/secrets.go:102]: not here since it's getter. added constructor

**Review by alexey-igrychev — APPROVED** [review]

**Pull request merged**